File: lightroom_tagger/core/analyzer/image_prep.py
"""Image preparation helpers (RAW/JPEG pipeline, compression, viewable paths)."""

import os
import tempfile

RAW_EXTENSIONS = {'.dng', '.raw', '.cr2', '.cr3', '.nef', '.arw', '.rw2', '.orf', '.raf', '.sr2', '.srw', '.x3f'}
VIDEO_EXTENSIONS = {'.mov', '.mp4', '.avi', '.mkv', '.wmv', '.m4v', '.3gp', '.webm', '.mts', '.m2ts'}

# Vision compression configuration
VISION_MAX_DIMENSION = int(os.environ.get('VISION_MAX_DIMENSION', '1024'))
VISION_COMPRESS_QUALITY = int(os.environ.get('VISION_COMPRESS_QUALITY', '80'))


def compress_image(
    input_path: str,
    max_size: tuple[int, int] | None = None,
    quality: int | None = None,
    *,
    silent: bool = False,
) -> str:
    """Compress image to reduce file size for vision comparison.

    Returns path to temporary compressed file.
    Caller is responsible for cleaning up the temporary file.

    Args:
        input_path: Path to input image
        max_size: Max (width, height) tuple, defaults to VISION_MAX_DIMENSION
        quality: JPEG quality (1-100), defaults to VISION_COMPRESS_QUALITY

    Returns:
        Path to compressed temporary file, or original path on failure.
    """
    from PIL import Image

    if max_size is None:
        max_size = (VISION_MAX_DIMENSION, VISION_MAX_DIMENSION)
    if quality is None:
        quality = VISION_COMPRESS_QUALITY

    try:
        with Image.open(input_path) as img:
            # Convert to RGB if necessary
            if img.mode in ('RGBA', 'LA', 'P'):
                img = img.convert('RGB')

            # Resize if larger than max_size
            if img.width > max_size[0] or img.height > max_size[1]:
                img.thumbnail(max_size, Image.Resampling.LANCZOS)

            # Save to temp file with compression
            fd, temp_path = tempfile.mkstemp(suffix='.jpg')
            os.close(fd)
            img.save(temp_path, 'JPEG', quality=quality, optimize=True)

            # Log compression
            original_size = os.path.getsize(input_path) / 1024 # KB
            compressed_size = os.path.getsize(temp_path) / 1024 # KB
            if not silent:
                print(f" Compressed: {original_size:.1f}KB -> {compressed_size:.1f}KB", flush=True)

            return temp_path
    except Exception as e:
        if not silent:
            print(f" Compression failed: {e}", flush=True)
        return input_path


def convert_raw_to_jpg(raw_path: str) -> str | None:
    """Convert RAW/DNG file to temporary JPG for vision comparison.

    Uses retry logic for network/NAS intermittent failures.

    Returns:
        Path to temporary JPG file, or None if conversion failed.
        Caller is responsible for cleaning up the temporary file.
    """
    import time

    import rawpy

    if not os.path.exists(raw_path):
        return None

    max_retries = 3
    for attempt in range(max_retries):
        try:
            with rawpy.imread(raw_path) as raw:
                rgb = raw.postprocess(use_camera_wb=True, half_size=True)

            from PIL import Image
            img = Image.fromarray(rgb)

            fd, jpg_path = tempfile.mkstemp(suffix='.jpg')
            os.close(fd)

            img.save(jpg_path, 'JPEG', quality=95)

            return jpg_path
        except rawpy._rawpy.LibRawIOError:
            # Network/NAS intermittent error - retry
            if attempt < max_retries - 1:
                time.sleep(0.5 * (attempt + 1)) # Exponential backoff
                continue
            return None
        except rawpy._rawpy.LibRawTooBigError:
            # Image too large - can't recover
            return None
        except Exception:
            # Other errors - don't retry
            return None

    return None


def get_viewable_path(image_path: str) -> str:
    """Get a viewable image path, converting RAW/DNG to temporary JPG if needed.

    Returns:
        Path to a viewable image (JPG/PNG).
        Returns original path if already viewable.
        Returns temporary JPG path if RAW/DNG and no sidecar exists (caller must clean up).
        Returns persistent sidecar JPG path if one exists next to the RAW (caller must NOT delete).

    Use ``get_viewable_path_managed`` when you need to know whether the returned path
    is a temporary file that should be cleaned up vs. a persistent on-disk sidecar.
    """
    path, _ = get_viewable_path_managed(image_path)
    return path


def get_viewable_path_managed(image_path: str) -> tuple[str, bool]:
    """Get a viewable image path with ownership information.

    Returns:
        (path, is_temp) where ``is_temp=True`` means the path is a mkstemp-created
        temporary file that the caller **must** clean up, and ``is_temp=False`` means
        the path is either the original or a persistent on-disk sidecar that must
        **not** be deleted by the caller.
    """
    ext = os.path.splitext(image_path)[1].lower()

    if ext not in RAW_EXTENSIONS:
        return image_path, False

    jpg_sidecar = image_path.rsplit('.', 1)[0] + '.JPG'
    if os.path.exists(jpg_sidecar):
        return jpg_sidecar, False  # persistent sidecar — do not delete

    jpg_sidecar_lower = image_path.rsplit('.', 1)[0] + '.jpg'
    if os.path.exists(jpg_sidecar_lower):
        return jpg_sidecar_lower, False  # persistent sidecar — do not delete

    converted = convert_raw_to_jpg(image_path)
    if converted:
        return converted, True  # mkstemp-created temp — caller must clean up

    return image_path, False

File: apps/visualizer/backend-ts/src/imaging/image-prep.ts
/**
 * Image preparation — viewable paths and vision compression.
 * Port of `core/analyzer/image_prep.py`.
 *
 * `compressImage` uses sharp rather than the Pillow-exact resampler in
 * `pil-resample.ts`, and that is a deliberate split. The Pillow port exists
 * because CLIP embeddings and phashes are *compared numerically* against 43,451
 * rows the Python backend wrote, so a one-level difference in a pixel changes a
 * stored value. This path feeds a JPEG to a language model, which will describe
 * the same photograph either way — so sharp's faster resize is the right tool,
 * and using the slow exact one here would buy nothing.
 */
import { mkdtemp, writeFile } from 'node:fs/promises';
import { existsSync, statSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import sharp from 'sharp';
import { convertRawToJpg, isRawPath } from './raw-decode.js';

export { RAW_EXTENSIONS, VIDEO_EXTENSIONS, isRawPath, isVideoPath } from './raw-decode.js';

/** Longest edge of the image handed to a vision model. */
export const VISION_MAX_DIMENSION = Number(process.env.VISION_MAX_DIMENSION ?? '1024');
export const VISION_COMPRESS_QUALITY = Number(process.env.VISION_COMPRESS_QUALITY ?? '80');

let tempDirPromise: Promise<string> | null = null;

/** One temp directory per process, created lazily. */
async function tempDir(): Promise<string> {
  tempDirPromise ??= mkdtemp(join(tmpdir(), 'lt-vision-'));
  return tempDirPromise;
}

let tempCounter = 0;

/** A unique temp path. The caller owns cleanup. */
export async function makeTempJpgPath(): Promise<string> {
  const dir = await tempDir();
  tempCounter += 1;
  return join(dir, `prep-${process.pid}-${tempCounter}.jpg`);
}

/**
 * Compress an image for a vision call. Returns the new path, or the input path
 * unchanged on failure.
 *
 * Returning the input rather than throwing is faithful and load-bearing in one
 * direction and a hazard in the other: it keeps a batch running past one odd
 * file, but it also means a `.mov` sails through to the provider, which is why
 * `description-service.ts` short-circuits video *before* getting here.
 */
export async function compressImage(
  inputPath: string,
  opts: { maxSize?: number; quality?: number; silent?: boolean } = {},
): Promise<string> {
  const maxSize = opts.maxSize ?? VISION_MAX_DIMENSION;
  const quality = opts.quality ?? VISION_COMPRESS_QUALITY;

  try {
    const image = sharp(inputPath, { failOn: 'none' });
    const meta = await image.metadata();

    const pipeline = image
      // 16-bit and floating-point samples cannot be written as JPEG. Pillow
      // rescaled the sample range by hand; sharp's `toColourspace('srgb')` plus
      // the JPEG encoder does the equivalent, and `removeAlpha` covers the
      // RGBA/LA/P modes Pillow converted explicitly.
      .removeAlpha()
      .toColourspace('srgb');

    // Only shrink. `withoutEnlargement` reproduces Pillow's `thumbnail`, which
    // is a no-op when the image already fits.
    if ((meta.width ?? 0) > maxSize || (meta.height ?? 0) > maxSize) {
      pipeline.resize({
        width: maxSize,
        height: maxSize,
        fit: 'inside',
        withoutEnlargement: true,
        kernel: 'lanczos3',
      });
    }

    const outPath = await makeTempJpgPath();
    await pipeline.jpeg({ quality, mozjpeg: true }).toFile(outPath);

    if (!opts.silent) {
      const kb = (p: string): string => (statSync(p).size / 1024).toFixed(1);
      process.stderr.write(` Compressed: ${kb(inputPath)}KB -> ${kb(outPath)}KB\n`);
    }
    return outPath;
  } catch (e) {
    if (!opts.silent) {
      process.stderr.write(
        ` Compression failed: ${e instanceof Error ? e.message : String(e)}\n`,
      );
    }
    return inputPath;
  }
}

/**
 * A viewable path plus whether it is ours to delete.
 *
 * `isTemp: false` means the original or a persistent on-disk sidecar, which the
 * caller must NOT unlink — deleting a user's `.JPG` next to their RAW would be
 * data loss. `isTemp: true` is a file we created.
 */
export interface ViewablePath {
  path: string;
  isTemp: boolean;
}

/**
 * Resolve a viewable image, converting RAW to a temporary JPEG when needed.
 *
 * A sidecar JPEG next to the RAW is preferred over decoding: it is already the
 * photographer's rendering, it costs nothing, and RAW decode is ~872 ms. Both
 * case variants are checked because Lightroom writes `.JPG` while cameras and
 * other tools write `.jpg`.
 */
export async function getViewablePathManaged(imagePath: string): Promise<ViewablePath> {
  if (!isRawPath(imagePath)) return { path: imagePath, isTemp: false };

  const stem = imagePath.slice(0, imagePath.lastIndexOf('.'));
  for (const sidecar of [`${stem}.JPG`, `${stem}.jpg`]) {
    if (existsSync(sidecar)) return { path: sidecar, isTemp: false };
  }

  const converted = await convertRawToJpg(imagePath, makeTempJpgPath);
  if (converted) return { path: converted, isTemp: true };

  // Decode failed: hand back the RAW. `compressImage` will fail on it and
  // return it unchanged, and the provider will reject it — a clear failure
  // rather than a silent skip.
  return { path: imagePath, isTemp: false };
}

/** The viewable path only, for callers that do not manage temp files. */
export async function getViewablePath(imagePath: string): Promise<string> {
  return (await getViewablePathManaged(imagePath)).path;
}

/** Write bytes to a fresh temp JPEG path. Used by tests and the cache builder. */
export async function writeTempJpg(data: Buffer): Promise<string> {
  const path = await makeTempJpgPath();
  await writeFile(path, data);
  return path;
}
